# Team assignment - Boston

## Description of the dataset

In [13]:
import numpy as np #optional
import pandas as pd
from pandas import DataFrame

### Bike rentals

Read data on bike rentals:

In [14]:
bike = pd.read_csv("boston_2016.csv") 

In [15]:
#for later use
#bike16 = pd.read_csv("boston_2016.csv")
#bike17 = pd.read_csv("boston_2017.csv")
#bike18 = pd.read_csv("boston_2018.csv")
#bike19 = pd.read_csv("boston_2019.csv")

In [16]:
bike.columns

Index(['start_time', 'end_time', 'start_station_id', 'end_station_id',
       'start_station_name', 'end_station_name', 'bike_id', 'user_type'],
      dtype='object')

In [17]:
bike.head(3)

start_time             end_time  start_station_id  end_station_id  \
0  2016-01-01 00:08:07  2016-01-01 00:40:26                36              19   
1  2016-01-01 00:08:19  2016-01-01 00:40:27                36              19   
2  2016-01-01 00:10:15  2016-01-01 00:40:29                36              19   

                         start_station_name         end_station_name  bike_id  \
0  Boston Public Library - 700 Boylston St.  Buswell St. at Park Dr.     1406   
1  Boston Public Library - 700 Boylston St.  Buswell St. at Park Dr.     1121   
2  Boston Public Library - 700 Boylston St.  Buswell St. at Park Dr.      474   

    user_type  
0  Subscriber  
1  Subscriber  
2  Subscriber

In [18]:
bike.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236203 entries, 0 to 1236202
Data columns (total 8 columns):
 #   Column              Non-Null Count    Dtype 
---  ------              --------------    ----- 
 0   start_time          1236203 non-null  object
 1   end_time            1236203 non-null  object
 2   start_station_id    1236203 non-null  int64 
 3   end_station_id      1236203 non-null  int64 
 4   start_station_name  1236203 non-null  object
 5   end_station_name    1236203 non-null  object
 6   bike_id             1236203 non-null  int64 
 7   user_type           1236203 non-null  object
dtypes: int64(3), object(5)
memory usage: 75.5+ MB


Number of entries (1236203) = Non-Null count for all colums, meaning we are not missing any values.

**Cleaning the datset:**

In [19]:
bike.isnull()

start_time  end_time  start_station_id  end_station_id  \
0             False     False             False           False   
1             False     False             False           False   
2             False     False             False           False   
3             False     False             False           False   
4             False     False             False           False   
...             ...       ...               ...             ...   
1236198       False     False             False           False   
1236199       False     False             False           False   
1236200       False     False             False           False   
1236201       False     False             False           False   
1236202       False     False             False           False   

         start_station_name  end_station_name  bike_id  user_type  
0                     False             False    False      False  
1                     False             False    False      False  
2                     False             False    False      False  
3                     False             False    False      False  
4                     False             False    False      False  
...                     ...               ...      ...        ...  
1236198               False             False    False      False  
1236199               False             False    False      False  
1236200               False             False    False      False  
1236201               False             False    False      False  
1236202               False             False    False      False  

[1236203 rows x 8 columns]

### Weather

Read weather data:

In [20]:
weatherhour = pd.read_csv("weather_hourly_boston.csv")

In [21]:
weatherhour.columns

Index(['date_time', 'max_temp', 'min_temp', 'precip'], dtype='object')

In [22]:
weatherhour.head(3)

date_time  max_temp  min_temp  precip
0  2015-01-02 01:00:00      -1.1      -1.1     0.0
1  2015-01-02 02:00:00      -1.1      -1.1     0.0
2  2015-01-02 03:00:00      -0.6      -0.6     0.0

In [23]:
weatherhour.describe()

max_temp      min_temp        precip
count  43354.000000  43354.000000  43356.000000
mean      11.456717     11.395922      0.095581
std       10.118606     10.114095      0.294019
min      -22.800000    -22.800000      0.000000
25%        3.300000      3.300000      0.000000
50%       11.700000     11.700000      0.000000
75%       20.000000     19.850000      0.000000
max       36.100000     36.100000      1.000000

In [24]:
weatherhour.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 43848 entries, 0 to 43847
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   date_time  43354 non-null  object 
 1   max_temp   43354 non-null  float64
 2   min_temp   43354 non-null  float64
 3   precip     43356 non-null  float64
dtypes: float64(3), object(1)
memory usage: 1.3+ MB


Number of entries (43848) != Non-Null count for all colums, meaning we are (!) missing values.

**Cleaning the datset:**

In [25]:
weatherhour.isnull()

date_time  max_temp  min_temp  precip
0          False     False     False   False
1          False     False     False   False
2          False     False     False   False
3          False     False     False   False
4          False     False     False   False
...          ...       ...       ...     ...
43843      False     False     False   False
43844      False     False     False   False
43845      False     False     False   False
43846      False     False     False   False
43847      False     False     False   False

[43848 rows x 4 columns]

In [26]:
#display entries with NaN values
NaN_for_attribute = weatherhour[weatherhour["date_time"].isna()] #insert any attribute
NaN_for_attribute

date_time  max_temp  min_temp  precip
98          NaN       NaN       NaN     NaN
145         NaN       NaN       NaN     NaN
304         NaN       NaN       NaN     NaN
395         NaN       NaN       NaN     NaN
588         NaN       NaN       NaN     NaN
...         ...       ...       ...     ...
43019       NaN       NaN       NaN     NaN
43428       NaN       NaN       NaN     NaN
43570       NaN       NaN       NaN     NaN
43594       NaN       NaN       NaN     NaN
43669       NaN       NaN       NaN     NaN

[494 rows x 4 columns]

Mostly, for given entries values are missing for all attributes.

**Imputation** <br>
To maintain the completeness of the time series dataset, we have decided to impute missing values.